docs/development/design/build-images.rst

```diff
@@ -0,0 +1,216 @@
+Besides, it proposes a new way to create and name them to allow
+sharing as many image layers as possible to support more customization while keeping the stability.
+
+
+Introduction
+------------
+
+We use Docker images to build user's documentation.
+Each time a build is triggered, one of our VMs picks the task
+and go through different steps:
+
+#. run some application code to spin up a Docker image into a container
+#. execute git inside the container to clone the repository
+#. analyze and parse files from the repository *outside* the container
+#. create the environment and install docs' dependencies inside the container
+#. execute build commnands inside the container
+#. push the output generated by builds commands to the storage
+
+
+*All* those steps depends on specific commands versions: ``git``, ``python``, ``virtualenv``, ``conda``, etc.
+Currently, we are pinning only a few of them in our Docker images and that have caused issues
+when re-deploying these images with bugfixes: **the images are not reproducible in time**.
+
+.. note::
+
+   The repoducibility of the images will be fixed once
+   https://github.com/readthedocs/readthedocs-docker-images/pull/145 and
+   https://github.com/readthedocs/readthedocs-docker-images/pull/146
+   get merged.
+
+To allow users to pin the image we ended up exposing three images: ``stable``, ``latest`` and ``testing``.
+With that naming, we were able to bugfix issues and add more features
+on each image without asking the users to change the image selected in their config file.
+
+Then, when a completely different image appeared and after testing ``testing`` image enough,
+we discarded ``stable``, old ``latest`` became the new ``stable`` and old ``testing`` became the new ``latest``.
+This produced issues to people pinning their images to any of these names because after this change,
+*we changed all the images for all the users* and many build issues arrised!
+
+
+Goals
+-----
+
+* release a completely new Docker image without forcing users to change their pinned image
+* allow users to stick with an image "forever" (~years)
+* use a ``base`` image with the dependencies that don't change frequently (OS and base requirements)
+* reduce size on builder VM disks by sharing Docker image layers
+* deprecate ``stable``, ``latest`` and ``testing``
+* allow use custom images for particular users/customers by sharing most layers
+* create a small ``nopdf`` image version without LaTeX dependencies for local development
+
+
+New build image structure
+-------------------------
+
+.. Taken from https://github.com/readthedocs/readthedocs-docker-images/blob/master/Dockerfile
+
+* ``ubuntu20-base``
+  * labels
+  * environment variables
+  * system dependencies
+  * install requirements
+  * user requirements
+    * plantuml, imagemagick, rsgv-convert, swig
+    * sphinx-js dependencies
+    * rust
+  * UID and GID
+
+* ``ubuntu20-pdf`` (from ``ubuntu20-base``)
```

Comment:
This should be called `ubuntu20-base-pdf` or something, otherwise these names are confusing.